net: l2: ieee802154: standardize RSSI value

fgrandel opened the pull request:
The RSSI value in net_pkt (net_pkt_cb_ieee802154.rssi) was used inconsistently across drivers. Some drivers did cast a signed dBm value directly to net_pkt's unsigned byte value. Others were assigning the negative value of the signed dBm value and again others were offsetting and stretching the signed dBm value linearly onto the full unsigned byte range.

This change standardizes net_pkt's rssi attribute to represent RSSI on the RX path as an unsigned integer ranging from 0 (–174 dBm) to 254 (80 dBm) and lets 255 represent an "unknown RSSI" (IEEE 802.15.4-2020, section 6.16.2.8). On the TX path the rssi attribute will always be zero. Out-of-range values will be truncated to max/min values.

The change also introduces conversion functions to and from signed dBm values and introduces these consistently to all existing call sites. The "unknown RSSI" value is represented as INT16_MIN in this case.

In some cases drivers had to be changed to calculate dBm values from internal hardware specific representations.

The conversion functions are fully covered by unit tests.

Fixes: #58494

Commit:
net: l2: ieee802154: standardize RSSI value

The RSSI value in net_pkt (net_pkt_cb_ieee802154.rssi) was used
inconsistently across drivers. Some drivers did cast a signed dBm value
directly to net_pkt's unsigned byte value. Others were assigning the
negative value of the signed dBm value and again others were offsetting
and stretching the signed dBm value linearly onto the full unsigned byte
range.

This change standardizes net_pkt's rssi attribute to represent RSSI on
the RX path as an unsigned integer ranging from 0 (–174 dBm) to 254 (80
dBm) and lets 255 represent an "unknown RSSI" (IEEE 802.15.4-2020,
section 6.16.2.8). On the TX path the rssi attribute will always be
zero. Out-of-range values will be truncated to max/min values.

The change also introduces conversion functions to and from signed dBm
values and introduces these consistently to all existing call sites. The
"unknown RSSI" value is represented as INT16_MIN in this case.

In some cases drivers had to be changed to calculate dBm values from
internal hardware specific representations.

The conversion functions are fully covered by unit tests.

Fixes: #58494

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

cfriedt commented on include/zephyr/net/ieee802154_pkt.h:
```diff
@@ -82,16 +95,89 @@ static inline void net_pkt_set_ieee802154_lqi(struct net_pkt *pkt, uint8_t lqi)
+ */
+static inline void net_pkt_set_ieee802154_rssi_dbm(struct net_pkt *pkt, int16_t rssi)
+{
+	if (likely(rssi >= -174 && rssi <= 80)) {
```
Can you please `#define` physical layer constants for `-174` and `80` and use those here?

fgrandel replied:
Fixed.

cfriedt commented on include/zephyr/net/ieee802154_pkt.h:
```diff
@@ -82,16 +95,89 @@ static inline void net_pkt_set_ieee802154_lqi(struct net_pkt *pkt, uint8_t lqi)
+		net_pkt_cb_ieee802154(pkt)->rssi = unsigned_rssi;
+		return;
+	} else if (rssi == INT16_MIN) {
+		net_pkt_cb_ieee802154(pkt)->rssi = UINT8_MAX;
```
Can you please also `#define` a physical layer constant for `UINT8_MAX`?

I'm also OK if this is defined as something like `((int8_t)-1)`, but I guess the spec says `0xff`?

fgrandel replied:
Fixed.

cfriedt commented on include/zephyr/net/ieee802154_pkt.h:
```diff
@@ -82,16 +95,89 @@ static inline void net_pkt_set_ieee802154_lqi(struct net_pkt *pkt, uint8_t lqi)
+	/* As the above rule covers the whole int16_t range,
+	 * this code should never be reached.
+	 */
+	__ASSERT_NO_MSG(false);
```
```suggestion
	CODE_UNREACHABLE;
```

fgrandel replied:
Fixed.

cfriedt commented on tests/net/ieee802154/l2/src/ieee802154_test.c:
```diff
@@ -535,6 +535,66 @@ static bool test_packet_cloning_with_cb(void)
+	zassert_equal(net_pkt_ieee802154_rssi(pkt), 50U);
+
+	/* Test setting/getting of signed RSSI (in range). */
+	net_pkt_set_ieee802154_rssi_dbm(pkt, -174);
```
Looks great - thanks for adding to the testsuite :-)

The only suggestion I have is to use `#define`ed constant suggested earlier.

fgrandel replied:
Fixed.

cfriedt reviewed (changes requested):
Just minor changes requested. Overall - wow - thanks @fgrandel for your work here. I think this is the sort of consistence we require for interfacing with the physical layer.

A suggestion as well, can you please add your GitHub UserID (and mine as well for that matter) under `collaborators` for `Drivers: IEEE 802.15.4`? As long as there are no objections.

Commit:
MAINTAINERS: add IEEE 802.15.4 driver collaborators

Added cfriedt and fgrandel as collaborators for IEEE 802.15.4 driver
maintenance as suggested by cfried in
https://github.com/zephyrproject-rtos/zephyr/pull/59133#pullrequestreview-1473704118

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

fgrandel commented:
> A suggestion as well, can you please add your GitHub UserID (and mine as well for that matter) under `collaborators` for `Drivers: IEEE 802.15.4`? As long as there are no objections.

Added a (tentative) commit as suggested by @cfriedt . Maintainer/reviewers - pls object and I'l remove the commit again.

cfriedt reviewed (dismissed):
LGTM in terms of common rssi, cc13xx_cc26xx, and MAINTAINERS changes

cfriedt commented:
Just adding a DNM for now until `v3.4.0` is tagged

fgrandel commented:
@cfriedt Inspired by your comments also fixed some magic RSSI values in CCxxxx drivers.

rlubos reviewed (approved):
LGTM
